gossipd/gossipd.c

```diff
@@ -341,6 +342,53 @@ static void handle_local_private_channel(struct daemon *daemon, const u8 *msg)
 	}
 }
 
+/* lightningd tells us it has dicovered and verified new `remote_addr`.
+ * We can use this to update our node announcement. */
+static void handle_remote_addr(struct daemon *daemon, const u8 *msg)
+{
+	struct wireaddr remote_addr;
+
+	if (!fromwire_gossipd_remote_addr(msg, &remote_addr))
+		master_badmsg(WIRE_GOSSIPD_REMOTE_ADDR, msg);
+
+	/* current best guess is that we use DEFAULT_PORT on public internet */
+	remote_addr.port = DEFAULT_PORT;
+
+	switch (remote_addr.type) {
+	case ADDR_TYPE_IPV4:
+		if (daemon->remote_addr_v4 != NULL &&
+		    wireaddr_eq_without_port(daemon->remote_addr_v4,
+					     &remote_addr))
+			break;
+		if (daemon->remote_addr_v4 == NULL)
+			daemon->remote_addr_v4 = tal(daemon, struct wireaddr);
+		*daemon->remote_addr_v4 = remote_addr;
```

Comment:
I prefer a "tal_free and tal_dup" pattern in general.  Sure, it's a bit wasteful to reallocate, but it's clear.